internal/server/singleprocess/service_workspace_test.go

```diff
@@ -0,0 +1,70 @@
+package singleprocess
+
+import (
+	"context"
+	"testing"
+
+	"github.com/hashicorp/waypoint/internal/server"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+	serverptypes "github.com/hashicorp/waypoint/internal/server/ptypes"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpsertWorkspace(t *testing.T) {
+	ctx := context.Background()
+
+	// Create our server
+	impl, err := New(WithDB(testDB(t)))
+	require.NoError(t, err)
+	client := server.TestServer(t, impl)
+
+	// Simplify writing tests
+	type Req = pb.UpsertWorkspaceRequest
+
+	t.Run("create and update", func(t *testing.T) {
+		require := require.New(t)
+
+		// Create
+		{
+			resp, err := client.UpsertWorkspace(ctx, &Req{
+				Workspace: serverptypes.TestWorkspace(t, &pb.Workspace{
+					Name: "staging",
+				}),
+			})
+			require.NoError(err)
+			require.NotNil(resp)
+		}
+
+		// Create another
```

Comment:
Can we also add a test for updating an existing workspace?

Reply:
At this time there is no real updating happening at all. At this time we don't yet support updating the workspace with an associated project (or removing one either), and we don't support renaming workspaces.

Reply:
To clarify, we absolutely should have tests on updating workspace and project associations, but I think a discussion needs to be had on if we support that through this Upsert method, or an other endpoint just for adding/removing projects.

Reply:
Oh, thanks for clarifying. Yeah, I was a bit confused reading through the code why this was an upsert instead of a create.